// @tetracyl/erela.js/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Utils_1 = require("./structures/Utils");
Object.defineProperty(exports, "Structure", { enumerable: true, get: function () { return Utils_1.Structure; } });
Object.defineProperty(exports, "Plugin", { enumerable: true, get: function () { return Utils_1.Plugin; } });
Object.defineProperty(exports, "LoadType", { enumerable: true, get: function () { return Utils_1.LoadType; } });
Object.defineProperty(exports, "State", { enumerable: true, get: function () { return Utils_1.State; } });
Object.defineProperty(exports, "Utils", { enumerable: true, get: function () { return Utils_1.Utils; } });
var Manager_1 = require("./structures/Manager");
Object.defineProperty(exports, "Manager", { enumerable: true, get: function () { return Manager_1.Manager; } });
var Player_1 = require("./structures/Player");
Object.defineProperty(exports, "Player", { enumerable: true, get: function () { return Player_1.Player; } });
var Queue_1 = require("./structures/Queue");
Object.defineProperty(exports, "Queue", { enumerable: true, get: function () { return Queue_1.Queue; } });
var Node_1 = require("./structures/Node");
Object.defineProperty(exports, "Node", { enumerable: true, get: function () { return Node_1.Node; } });
